File: backend/app/api/categories.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from backend.app.database.database import SessionLocal
from backend.app.models.category import Category
from backend.app.schemas.category import (
    CategoryCreate,
    CategoryResponse,
    CategoryUpdate,
)


router = APIRouter(
    prefix="/api/categories",
    tags=["Categories"],
)


def get_db():
    db = SessionLocal()

    try:
        yield db
    finally:
        db.close()


@router.post(
    "",
    response_model=CategoryResponse,
    status_code=201,
)
def create_category(
    category: CategoryCreate,
    db: Session = Depends(get_db),
):

    existing_category = (
        db.query(Category)
        .filter(Category.name == category.name)
        .first()
    )

    if existing_category:
        raise HTTPException(
            status_code=400,
            detail="Category already exists",
        )

    new_category = Category(
        name=category.name,
        description=category.description,
        priority=category.priority,
        color=category.color,
        icon=category.icon,
        is_active=category.is_active,
    )

    db.add(new_category)
    db.commit()
    db.refresh(new_category)

    return new_category


@router.get(
    "",
    response_model=list[CategoryResponse],
)
def get_categories(
    db: Session = Depends(get_db),
):

    return (
        db.query(Category)
        .order_by(
            Category.priority.desc(),
            Category.name.asc(),
        )
        .all()
    )


@router.get(
    "/{category_id}",
    response_model=CategoryResponse,
)
def get_category(
    category_id: int,
    db: Session = Depends(get_db),
):

    category = (
        db.query(Category)
        .filter(Category.id == category_id)
        .first()
    )

    if not category:
        raise HTTPException(
            status_code=404,
            detail="Category not found",
        )

    return category


@router.put(
    "/{category_id}",
    response_model=CategoryResponse,
)
def update_category(
    category_id: int,
    category_data: CategoryUpdate,
    db: Session = Depends(get_db),
):

    category = (
        db.query(Category)
        .filter(Category.id == category_id)
        .first()
    )

    if not category:
        raise HTTPException(
            status_code=404,
            detail="Category not found",
        )

    update_data = category_data.model_dump(
        exclude_unset=True,
    )

    if "name" in update_data:

        existing_category = (
            db.query(Category)
            .filter(
                Category.name == update_data["name"],
                Category.id != category_id,
            )
            .first()
        )

        if existing_category:
            raise HTTPException(
                status_code=400,
                detail="Category already exists",
            )

    for field, value in update_data.items():
        setattr(category, field, value)

    db.commit()
    db.refresh(category)

    return category


@router.delete(
    "/{category_id}",
    status_code=204,
)
def delete_category(
    category_id: int,
    db: Session = Depends(get_db),
):

    category = (
        db.query(Category)
        .filter(Category.id == category_id)
        .first()
    )

    if not category:
        raise HTTPException(
            status_code=404,
            detail="Category not found",
        )

    db.delete(category)
    db.commit()

File: frontend/src/api/categories.ts
import client from "./client"


export interface Category {
    id: number
    name: string
    description: string | null
    priority: number
    color: string
    icon: string | null
    is_active: boolean
    created_at: string
}


export interface CreateCategoryData {
    name: string
    description?: string | null
    priority: number
    color: string
    icon?: string | null
    is_active: boolean
}


export interface UpdateCategoryData {
    name?: string
    description?: string | null
    priority?: number
    color?: string
    icon?: string | null
    is_active?: boolean
}


export async function getCategories(): Promise<Category[]> {

    const response = await client.get(
        "/categories"
    )

    return response.data
}


export async function createCategory(
    data: CreateCategoryData
): Promise<Category> {

    const response = await client.post(
        "/categories",
        data
    )

    return response.data
}


export async function updateCategory(
    categoryId: number,
    data: UpdateCategoryData
): Promise<Category> {

    const response = await client.put(
        `/categories/${categoryId}`,
        data
    )

    return response.data
}


export async function deleteCategory(
    categoryId: number
): Promise<void> {

    await client.delete(
        `/categories/${categoryId}`
    )
}